Feature/apt 944 rvc udp packet

psilar opened the pull request:
This code is not ready for merging to develop yet.  Please take a quick look at the code/API and we can discuss it in the PG or another meeting later, before going through the PR review process.

Commit:
added draft of Radar EGO Motion (RVC) UDP packet spec for discussion

Commit:
changed radar_ego_motion structures to agreed format after meeting with Peter, Iurii & Dane

Commit:
Draft implementation of support for EGO motion UDP packet type
 - moved provizio_radar_position enum from radar_point_cloud.h to common.h
 - added radar_ego_motion.c to provizio_radar_api_core library
 - handling of contexts may need restructuring (TBD)
 - no unit test coverage

Commit:
undoinging radar_position change pending rebase'ing off develop branch

Commit:
added static assertions for radar ego motion packet

iurii-provizio commented:
@psilar I finally reviewed it (understanding it's not ready for merging of course).
Let's indeed have a call to discuss, but it's done not the way I would have expected it to be done.

I see those ego motion packets (which are themselves alright, except some minor changes I'd suggest) as an internal thing - the API customer doesn't have to ever deal with them. Instead, they should be handled by `provizio_radar_api_receive_packet` and use same `provizio_radar_point_cloud_api_context` structures internally as the point clouds. Then on receiveing those packets, their appropriate `vs_x` and `vs_y` would be internally stored in those context(s), and used to estimate ground-relative velocities for point clouds when they're later received, adding an additional "ground relative velocity" float field to `provizio_radar_point`.

psilar commented:
@iurii-provizio I like the idea of adding a ground_velocity field to `provizio_radar_point` for ease of use, but think the API should also support an "advanced" mode where users can access the `vs_x` and `vs_y` values directly.  The impetus for this would be allowing the user to use these values in conjunction with other inputs, e.g. vehicle telematics, to better calculate sensor velocity / magnitude before applying it to a point cloud.  We can discuss this on the call.

Commit:
Work in progress - committed for Iurii to review
 - refactored api_context into its own file (common to point_cloud and ego_motion)
 - provizio_radar_point: changed velocity_m_s to radar_relative_radial_velocity_m_s
 - provizio_radar_point: added ground_relative_radial_velocity_m_s

Commit:
text fix

Commit:
fixed ego motion packet type macro definition

iurii-provizio commented on include/provizio/radar_api/radar_api_context.h:
```diff
@@ -0,0 +1,98 @@
+#include "provizio/radar_api/radar_point_cloud.h"
+
+// To be incremented on any breaking protocol changes (used for backward compatibility)
+#define PROVIZIO__RADAR_API_CONTEXT_PROTOCOL_VERSION ((uint16_t)1)
```
@psilar It looks wrong. I.e. there are protocol versions which define UDP packets structure. API context has nothing to do with UDP packets really, it's memory storage for API. There should be separate protocol versions defined for point clouds packets, and seperate one for ego motion packets. Separate versions on purpose - so in case of updating protocol, everything that didn't change the format stays forward compatible.

psilar replied:
Makes sense .. will remove that macro.

iurii-provizio commented on include/provizio/radar_api/radar_api_context.h:
```diff
@@ -0,0 +1,98 @@
+    uint16_t radar_position_id;
+    provizio_radar_point_cloud_callback point_cloud_callback;
+    void *point_cloud_user_data;
+    provizio_radar_ego_motion_callback ego_motion_callback;
```
@psilar So you decided to add this callback after all? OK, if you think it's of any use, I'm OK with it.

psilar replied:
Because there in no guarantee of an ego motion packet arriving 1) at all 2) in order 3) in time before the related point cloud has been processed, I think it's prudent to at least allow a user to explicitly handle it.  I'll revisit this after integrating ego motion with accumulation.

iurii-provizio commented on include/provizio/radar_api/radar_ego_motion.h:
```diff
@@ -0,0 +1,178 @@
+ * @see provizio_set_protocol_field_uint16_t
+ * @see provizio_get_protocol_field_uint16_t
+ */
+typedef struct provizio_radar_ego_motion_packet_protocol_header
```
@psilar There is now a common definition for this structure. It's not per packet type. It's probably in common.h

psilar replied:
Ok .. will use common definition

iurii-provizio commented on include/provizio/radar_api/radar_ego_motion.h:
```diff
@@ -0,0 +1,178 @@
+#pragma pack(push, 1)
+
+// declared in a radar_api_context.h which is not included to avoid circular dependencies
+struct provizio_radar_api_context;
```
@psilar Looks a bit weird that it's declared under the `#pragma pack` block

psilar replied:
Agreed .. will move it outside

iurii-provizio commented on include/provizio/radar_api/radar_ego_motion.h:
```diff
@@ -0,0 +1,178 @@
+    provizio_radar_ego_motion_packet_protocol_header protocol_header;
+
+    uint32_t frame_index;   // 0-based
+    uint64_t timestamp;     // Time of the frame capture measured in absolute number of nanoseconds since the start of the
```
@psilar Please timestamp should be above frame_index as it should be 64-bits aligned (otherwise reading/writing operations are much slower, though safe).

psilar replied:
Good catch

psilar replied:
Was about to change this but realized the timestamp is already 64-bit aligned because the protocol_header size is four bytes.

iurii-provizio commented on include/provizio/radar_api/radar_ego_motion.h:
```diff
@@ -0,0 +1,178 @@
+                            // GPS Epoch (midnight on Jan 6, 1980)
+    uint16_t radar_position_id;   // Either one of provizio_radar_position enum values or a custom position id
+    uint16_t reserved;      // Not used currently, kept for better alignment and potential future use
+    float vs_x;             // sensor ego motion x
```
@psilar I normally try to keep names in the protocol and API as descriptive as possible to avoid any potential confusion. `vs_x` is a type of name DSP team likes - short and mathematical, but I'd rather use something that our customers would understand (given they have that callback to deal with this thing). I.e. what it stands for, and what units does it measure (m/s or something else?).

psilar replied:
In my head it's a 2D vector from 0,0 to x,y representing the sensor velocity but I'll check with Peter to find an accurate description of it.

psilar replied:
Are you happy with sensor_velocity_x _m_s and sensor_velocity_y_m_s ?

iurii-provizio replied:
@psilar I normally use "radar" instead of "sensor" in the API - it's the Radar API after all, not some other sensor API. Plus, given they're in m/s, I'd call them `radar_velocity_x_m_s` and `radar_velocity_y_m_s`.

iurii-provizio commented on include/provizio/radar_api/radar_ego_motion.h:
```diff
@@ -0,0 +1,178 @@
+ */
+typedef struct provizio_radar_ego_motion
+{
+    uint32_t frame_index;   // 0-based
```
Though less important than in packed structures, here I'd reordered the fields too (otherwise you're introducing a 4 bytes gap without any reason).

psilar replied:
Agreed   .. will move timestamp to top

iurii-provizio commented on include/provizio/radar_api/radar_ego_motion.h:
```diff
@@ -0,0 +1,178 @@
+    uint32_t frame_index;   // 0-based
+    uint64_t timestamp;     // Time of the frame capture measured in absolute number of nanoseconds since the start of the
+                            // GPS Epoch (midnight on Jan 6, 1980)
+    float vs_x;             // sensor ego motion x
```
Same comments as above regarding naming for `vs_x`

iurii-provizio commented on include/provizio/radar_api/radar_point_cloud.h:
```diff
@@ -28,6 +28,9 @@
 #pragma pack(push, 1)
 
+// declared in a radar_api_context.h which is not included to avoid circular dependencies
+struct provizio_radar_api_context;
```
@psilar Same thing as before about declaration under `#pragma pack(push, 1)`

psilar replied:
Will move outside

iurii-provizio commented on include/provizio/radar_api/radar_point_cloud.h:
```diff
@@ -41,7 +44,8 @@ typedef struct
     float z_meters;     // Up, radar relative
-    float velocity_m_s; // Forward, radar relative
+    float radar_relative_radial_velocity_m_s;   // Forward, radar relative
+    float ground_relative_radial_velocity_m_s;  // Forward, ground relative
```
@psilar Comment is not very descriptive here. Radar does have a forward direction. Ground doesn't. So it's worth noting that though the value is ground relative, it's in projection to the forward axis of the radar.

psilar replied:
What would you suggest?  Would "Ground relative" be sufficient?

iurii-provizio replied:
@psilar Maybe, `Ground relative projection to the radar forward axis` or something like that to deliver this important idea that's it's not the entire velocity, but only it's radar-radial component.

psilar replied:
Comment updated.

iurii-provizio commented on include/provizio/radar_api/radar_point_cloud.h:
```diff
@@ -119,98 +123,50 @@ typedef struct provizio_radar_point_cloud
     uint16_t num_points_expected; // Number of points in the entire frame
     uint16_t num_points_received; // Number of points in the frame received so far
     uint16_t radar_mode;          // One of provizio_radar_mode enum values
+    float vs_x;                   // EGO velocity X for this frame (can be NaN)
```
@psilar I'd suggest `can be NaN when unknown`.

psilar replied:
Comment updated. Thought I might be useful to have vs_xy here pending integrating with accumulation api.  TBD.

iurii-provizio commented on include/provizio/radar_api/radar_point_cloud.h:
```diff
@@ -119,98 +123,50 @@ typedef struct provizio_radar_point_cloud
-#endif // PROVIZIO__RADAR_POINT_CLOUD_API_CONTEXT_IMPL_POINT_CLOUDS_BEING_RECEIVED_COUNT
-typedef struct provizio_radar_point_cloud_api_context_impl
+#ifndef PROVIZIO__RADAR_POINT_CLOUD_IMPL_POINT_CLOUDS_BEING_RECEIVED_COUNT
+#define PROVIZIO__RADAR_POINT_CLOUD_IMPL_POINT_CLOUDS_BEING_RECEIVED_COUNT 2
```
@psilar It's not point cloud impl, it's exactly the api context impl. So it's now rather `PROVIZIO__RADAR_API_CONTEXT_IMPL_POINT_CLOUDS_BEING_RECEIVED_COUNT`

iurii-provizio commented on src/radar_api_context.c:
```diff
@@ -0,0 +1,116 @@
+    context->ego_motion_callback = ego_motion_callback;
+    context->ego_motion_user_data = ego_motion_user_data;
+    context->radar_position_id = provizio_radar_position_unknown;
+    context->motion.vs_x = NAN;
```
@psilar `NAN` macro is not guaranteed by C99 standard (though may be present). Instead there is a guaranteed function `nanf`. See https://stackoverflow.com/a/1923933 and https://www.ibm.com/docs/en/zos/2.2.0?topic=functions-nan-nanf-nanl-return-quiet-nan. You may safely use `nanf("")`.

psilar replied:
Fixed

iurii-provizio commented on src/radar_api_context.c:
```diff
@@ -0,0 +1,116 @@
+    context->motion.vs_y = NAN;
+    for (size_t i = 0; i < PROVIZIO__RADAR_POINT_CLOUD_IMPL_POINT_CLOUDS_BEING_RECEIVED_COUNT; ++i)
+    {
+        context->impl.point_clouds_being_received[i].vs_x = NAN;
```
@psilar but really, why do you need vs_x/vs_y in both the context and point clouds? Am I missing some important idea?

psilar replied:
The context only stores the most recent vs_xy and when considering udp packets being out-of-order and/or delayed I thought it might be useful to save vs_xy values with their associated point clouds pending calculation of ground velocity. Another option might be to save the X most recent motion values.  TBD.

iurii-provizio replied:
@psilar Still I don't see any need in these values in point clouds anyway, as long as corrected ground-relative velocities are available in every point. Accumulation would rely on ground relative velocities when they're available, not vs_x/vs_y directly.

iurii-provizio commented on src/radar_points_accumulation.c:
```diff
@@ -65,7 +65,7 @@ static void provizio_transform_radar_point(const provizio_radar_point *point, co
     out_transformed_point->y_meters = out_point[1];
     out_transformed_point->z_meters = out_point[2];
-    out_transformed_point->velocity_m_s = point->velocity_m_s; // TODO(APT-746): Consider updating velocity
+    out_transformed_point->radar_relative_radial_velocity_m_s = point->radar_relative_radial_velocity_m_s; // TODO(APT-746): Consider updating velocity
```
@psilar what about ground relative velocity?

psilar replied:
Yes.  Haven't gone through the accumulation code yet other than fix compile errors.  Will be focusing on it next.

iurii-provizio commented on src/radar_points_accumulation_filters.c:
```diff
@@ -44,7 +44,7 @@ static float provizio_estimate_radars_forward_velocity_using_velocities_histogra
 
     for (uint16_t i = 0; i < num_in_points; ++i)
```
@psilar The filter also has to be updated: if ground relative velocities are available, then no need to estimate them using localization. Otherwise as currently.

psilar replied:
Will update the filter too.

Commit:
fixes as suggested by Iurii from code review

Commit:
Renamed vs_x/y to sensor_velocity_x/y
Substituted NAN with nanf() or isnan() as appropriate (C99)

Commit:
Rearranged ego motion packet structure for better alignment

Commit:
fixes from code review

Commit:
changed provizio_radar_point to match apt-1617
